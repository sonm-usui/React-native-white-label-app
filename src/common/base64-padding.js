// @flow

export function addBase64Padding(data: string): string {
  const trail = data.length % 4
  if (trail === 0) {
    return data
  }

  return `${data}${'='.repeat(4 - trail)}`
}

export function removeBase64Padding(data: string): string {
  return data.split('=')[0]
}
