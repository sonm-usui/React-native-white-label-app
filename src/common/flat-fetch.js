// @flow

// If you look into the return type of this function
// you should notice that we are returning an array from  this function
// this function is just a wrapper on top of fetch.
// However, this changes a lot of ways we write code and handle errors
// By returning an array in which first item is an error
// and second item is the response
// We get two important things
// 1. We don't have to write try...catch statements everywhere
// 2. We are forced to handle errors. No more forgetting to add try...catch
// TODO:KS Write more details about both 1 & 2
// Problem #1 leads to try...catch nesting hell if there are multiple async
// calls
// Problem #2 leads to unexpected app crashes at runtime

export const acceptHeaders = { Accept: 'application/json;flavor=didcomm-msg' }

export async function flatFetch(
  url: string,
  body?: string,
  headers?: { [string]: any }
): Promise<[null | typeof Error, null | string, null | Response]> {
  const options = {
    mode: 'cors',
    ...(body ? { method: 'POST', body } : {}),
    ...(headers ? headers : {}),
  }
  try {
    const response = await fetch(url, options)
    const responseText = await response.text()
    // request redirected
    if (response && response.status === 302) {
      return [null, null, response]
    }
    if (!response.ok) {
      throw new Error(responseText)
    }
    return [null, responseText, response]
  } catch (e) {
    return [e, null, null]
  }
}
