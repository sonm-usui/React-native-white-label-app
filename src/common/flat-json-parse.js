// @flow

export function flatJsonParse(
  text: string
): [null | typeof Error, null | Object] {
  try {
    return [null, JSON.parse(text)]
  } catch (e) {
    return [e, null]
  }
}
