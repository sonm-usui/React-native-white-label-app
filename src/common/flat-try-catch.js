// @flow

export const flatTryCatch = (fn: (...variableArgs: any[]) => any) => (
  ...args: any[]
): [null | typeof Error, null | any] => {
  try {
    return [null, fn(...args)]
  } catch (e) {
    return [e, null]
  }
}
