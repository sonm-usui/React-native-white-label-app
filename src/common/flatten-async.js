// @flow

// If you look into the return type of this function
// you should notice that we are returning an array from  this function
// this is just a helper function to wrap on top of other async functions
// However, this changes a lot of ways we write code and handle errors
// By returning an array in which first item is an error
// and second item is the response
// We get two important things
// 1. We don't have to write try...catch statements everywhere
// 2. We are forced to handle errors. No more forgetting to add try...catch
// TODO:KS Write more details about both 1 & 2
// Here are the problems that this function aims to solve
// Problem #1 leads to try...catch nesting hell if there are multiple async
// calls
// Problem #2 leads to unexpected app crashes at runtime

export const flattenAsync = (
  fn: (...variableArgs: any[]) => Promise<any>
) => async (
  ...args: Array<any>
): Promise<[null | typeof Error, null | any]> => {
  try {
    const response = await fn(...args)
    return [null, response]
  } catch (e) {
    return [e, null]
  }
}
