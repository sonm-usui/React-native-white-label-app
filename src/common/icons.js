// @flow
import React from 'react'
import { Icon } from 'react-native-eva-icons'
import { colors } from './styles/constant'
import { moderateScale } from 'react-native-size-matters'

type IconProps = {
  width?: number,
  height?: number,
  color?: string,
}

type CommonIconProps = {
  name: string,
} & IconProps

export const HOME_ICON = 'home-outline'
export const CONNECTIONS_ICON = 'people-outline'
export const SETTINGS_ICON = 'settings-2-outline'

export const ERROR_ICON = 'alert-circle-outline'
export const CAMERA_ICON = 'camera-outline'
export const CHECK_MARK_ICON = 'checkmark-circle-2-outline'
export const HOME_MENU_ICON = 'menu-outline'
export const ANDROID_BACK_ARROW_ICON = 'arrow-back-outline'
export const IOS_BACK_ARROW_ICON = 'arrow-ios-back-outline'
export const MORE_ICON = 'more-vertical-outline'
export const DELETE_ICON = 'trash-2-outline'
export const CLOSE_ICON = 'close-outline'
export const CHAT_ICON = 'message-square-outline'
export const INFO_ICON = 'info-outline'
export const ARROW_RIGHT_ICON = 'chevron-right-outline'
export const LOCK_ICON = 'lock'
export const SAVE_ICON = 'save-outline'
export const ALERT_ICON = 'alert-circle-outline'
export const ARROW_FORWARD_ICON = 'arrow-ios-forward-outline'
export const CHECKMARK_ICON = 'checkmark-outline'
export const SHARE_ICON = 'share-outline'
export const GRID_ICON = 'grid-outline'

// common icon class is implemented to set default values (except name) for icons
// in that case we should only set icon name and get an icon with default color and size
export const EvaIcon = (props: CommonIconProps) => {
  const { width, height, color } = props

  return (
    <Icon
      {...props}
      width={width ? width : moderateScale(24)}
      height={height ? height : moderateScale(24)}
      fill={color ? color : colors.gray2}
    />
  )
}
