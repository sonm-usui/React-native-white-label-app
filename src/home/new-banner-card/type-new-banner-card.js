// @flow
import { removeEvent } from '../../connection-history/connection-history-store'
import type { ConnectionHistoryEvent } from '../../connection-history/type-connection-history'

export type NewBannerCardProps = {
  navigation: Object,
  navigationRoute: string,
  timestamp: string,
  logoUrl: string,
  uid: string,
  issuerName: string,
  removeEvent: typeof removeEvent,
}
