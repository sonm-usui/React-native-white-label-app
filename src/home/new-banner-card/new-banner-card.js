// @flow
import React, { useCallback } from 'react'
import { TouchableOpacity, Text, View, Image, StyleSheet } from 'react-native'
import { isiPhone5 } from '../../common/styles'
import {
  colors,
  fontFamily,
  fontSizes as fonts,
  fontSizes,
} from '../../common/styles/constant'
import { scale, moderateScale } from 'react-native-size-matters'

import type { NewBannerCardProps } from './type-new-banner-card'
import { DefaultLogo } from '../../components/default-logo/default-logo'
import { SwipeRow } from 'react-native-swipe-list-view'
import { removeEvent } from '../../connection-history/connection-history-store'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { renderUserAvatar } from "../../components/user-avatar/user-avatar";
import { EventTimestamp } from '../../components/event-timestamp/event-timestamp'

const NewBannerCardComponent = (props: NewBannerCardProps) => {
  const onDelete = useCallback(() => {
    props.removeEvent(props.uid, props.navigationRoute)
  }, [props.navigationRoute, props.uid])

  return (
    <SwipeRow rightOpenValue={-scale(75)}>
      <View style={styles.rowBack}>
        <TouchableOpacity style={styles.deleteButton} onPress={onDelete}>
          <Text style={styles.deleteButtonText}>Delete</Text>
        </TouchableOpacity>
      </View>
      <View style={styles.rowFront}>
        <TouchableOpacity
          style={styles.cardContainer}
          testID="new-message"
          accessible={true}
          accessibilityLabel="new-message"
          onPress={() =>
            props.navigation.navigate(props.navigationRoute, { uid: props.uid })
          }
        >
          <View style={styles.iconSection}>
            {typeof props.logoUrl === 'string' ? (
              <Image
                source={{ uri: props.logoUrl }}
                style={styles.issuerLogo}
              />
            ) :
              props.issuerName ?
              <DefaultLogo
                text={props.issuerName[0]}
                size={moderateScale(34, 0.15)}
                fontSize={isiPhone5 ? fonts.size4 : fonts.size3}
              /> :
                renderUserAvatar({ size: 'small' })
            }
          </View>
          <View style={styles.textSection}>
            <View style={styles.textIssuerSection}>
              <Text
                style={styles.issuerText}
                ellipsizeMode="tail"
                numberOfLines={1}
              >
                {props.issuerName || 'Unnamed Connection'}
              </Text>
            </View>
            <View style={styles.textMessageSection}>
              <Text style={styles.newMessageText}>
                NEW MESSAGE - TAP TO OPEN
              </Text>
            </View>
          </View>
          <View style={styles.textDateSection}>
            <EventTimestamp timestamp={props.timestamp} customStyles={styles.dateText}/>
          </View>
        </TouchableOpacity>
      </View>
    </SwipeRow>
  )
}

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      removeEvent,
    },
    dispatch
  )

export const NewBannerCard = connect(
  null,
  mapDispatchToProps
)(NewBannerCardComponent)

const styles = StyleSheet.create({
  rowFront: {
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: colors.white,
    marginLeft: moderateScale(7, 0.1),
    marginRight: moderateScale(7, 0.1),
    marginTop: moderateScale(7, 0.1),
  },
  rowBack: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    backgroundColor: colors.white,
    marginLeft: moderateScale(7, 0.1),
    marginRight: moderateScale(7, 0.1),
    marginTop: moderateScale(7, 0.1),
  },
  cardContainer: {
    flexDirection: 'row',
    backgroundColor: colors.secondary,
    borderWidth: 1,
    borderColor: colors.main,
    height: moderateScale(70, 0.12),
    borderRadius: 8,
    alignItems: 'center',
    justifyContent: 'center',
  },
  iconSection: {
    height: '100%',
    width: moderateScale(65, 0.2),
    alignItems: 'center',
    justifyContent: 'center',
  },
  textSection: {
    flex: 1,
  },
  textIssuerSection: {
    flex: 2,
    justifyContent: 'flex-end',
  },
  textMessageSection: {
    flex: 2,
    justifyContent: 'flex-start',
  },
  textDateSection: {
    height: '100%',
    width: moderateScale(65, 0.2),
    alignItems: 'flex-end',
    justifyContent: 'flex-start',
  },
  issuerText: {
    fontFamily: fontFamily,
    fontSize: moderateScale(fontSizes.size5, 0.1),
    fontWeight: 'bold',
    color: colors.gray1,
    marginBottom: moderateScale(3, 0.1),
  },
  newMessageText: {
    fontFamily: fontFamily,
    fontSize: moderateScale(fontSizes.size8, 0.1),
    fontWeight: 'bold',
    color: colors.gray1,
    marginTop: moderateScale(3, 0.1),
  },
  issuerLogo: {
    width: moderateScale(34, 0.15),
    height: moderateScale(34, 0.15),
    borderRadius: moderateScale(34, 0.15) / 2,
  },
  dateText: {
    fontFamily: fontFamily,
    fontWeight: '500',
    fontSize: moderateScale(fontSizes.size9, 0.1),
    fontStyle: 'italic',
    color: colors.gray1,
    marginTop: moderateScale(8, 0.1),
    marginRight: moderateScale(8, 0.1),
  },
  placeholderIfNoImage: {
    width: moderateScale(34, 0.15),
    height: moderateScale(34, 0.15),
    borderRadius: moderateScale(34, 0.15) / 2,
    backgroundColor: colors.gray2,
    alignItems: 'center',
    justifyContent: 'center',
  },
  placeholderTextIfNoImage: {
    fontFamily: fontFamily,
    fontSize: moderateScale(fontSizes.size4, 0.1),
    fontWeight: 'bold',
    color: colors.white,
  },
  deleteButton: {
    alignItems: 'center',
    justifyContent: 'center',
    position: 'absolute',
    bottom: 0,
    top: 0,
    right: 0,
    width: moderateScale(75, 0.12),
    height: moderateScale(70, 0.12),
    borderRadius: 8,
    backgroundColor: colors.red,
  },
  deleteButtonText: {
    color: colors.white,
    alignItems: 'center',
    fontFamily: fontFamily,
    fontSize: scale(14),
  },
})
