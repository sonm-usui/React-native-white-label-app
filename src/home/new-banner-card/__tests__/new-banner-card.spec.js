// @flow
import React from 'react'
import renderer from 'react-test-renderer'
import 'react-native'

import { getNavigation, getStore } from '../../../../__mocks__/static-data'
import { NewBannerCard } from '../new-banner-card'
import { claimOfferRoute } from '../../../common'
import { Provider } from 'react-redux'

describe('<NewBannerCard />', () => {
  const navigation = getNavigation()
  const store = getStore()

  it('should render NewBannerCard component and match snapshot', () => {
    const component = renderer
      .create(
        <Provider store={store}>
          <NewBannerCard
            navigation={navigation}
            navigationRoute={claimOfferRoute}
            timestamp="2020-03-15T15:57:54+01:00"
            logoUrl="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSjimxZbcvZzPcPvHd_y7f0tc5d8QoC9DOPecb8JTOChmS1IoDq"
            uid="MGEwNzg"
            issuerName="Evernym QA-RC"
          />
        </Provider>
      )
      .toJSON()
    expect(component).toMatchSnapshot()
  })
})
