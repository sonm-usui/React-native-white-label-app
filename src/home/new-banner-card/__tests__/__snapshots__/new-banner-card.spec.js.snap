// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewBannerCard /> should render NewBannerCard component and match snapshot 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "overflow": "hidden",
          "position": "absolute",
          "right": 0,
          "top": 0,
          "zIndex": 1,
        },
        Object {
          "height": 0,
          "width": 0,
        },
      ]
    }
  >
    <View
      leftActionActivated={false}
      leftActionState={false}
      rightActionActivated={false}
      rightActionState={false}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginLeft": 6.9399999999999995,
          "marginRight": 6.9399999999999995,
          "marginTop": 6.9399999999999995,
        }
      }
      swipeAnimatedValue={0}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#CE0B24",
            "borderRadius": 8,
            "bottom": 0,
            "height": 69.28,
            "justifyContent": "center",
            "opacity": 1,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 74.22857142857143,
          }
        }
      >
        <Text
          style={
            Object {
              "alignItems": "center",
              "color": "#FFFFFF",
              "fontFamily": "System",
              "fontSize": 12.799999999999999,
            }
          }
        >
          Delete
        </Text>
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    manipulationModes={
      Array [
        "translateX",
      ]
    }
    nativeID="animatedComponent"
    onLayout={[Function]}
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      Object {
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
        "zIndex": 2,
      }
    }
  >
    <View
      accessible={false}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        leftActionActivated={false}
        leftActionState={false}
        rightActionActivated={false}
        rightActionState={false}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "justifyContent": "center",
            "marginLeft": 6.9399999999999995,
            "marginRight": 6.9399999999999995,
            "marginTop": 6.9399999999999995,
          }
        }
        swipeAnimatedValue={0}
      >
        <View
          accessibilityLabel="new-message"
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(134, 185, 59, 0.15)",
              "borderColor": "#86B93B",
              "borderRadius": 8,
              "borderWidth": 1,
              "flexDirection": "row",
              "height": 69.28,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
          testID="new-message"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "height": "100%",
                "justifyContent": "center",
                "width": 63.885714285714286,
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSjimxZbcvZzPcPvHd_y7f0tc5d8QoC9DOPecb8JTOChmS1IoDq",
                }
              }
              style={
                Object {
                  "borderRadius": 16.78142857142857,
                  "height": 33.56285714285714,
                  "width": 33.56285714285714,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 2,
                  "justifyContent": "flex-end",
                }
              }
            >
              <Text
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  Object {
                    "color": "#505050",
                    "fontFamily": "System",
                    "fontSize": 14.871428571428572,
                    "fontWeight": "bold",
                    "marginBottom": 2.974285714285714,
                  }
                }
              >
                Evernym QA-RC
              </Text>
            </View>
            <View
              style={
                Object {
                  "flex": 2,
                  "justifyContent": "flex-start",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#505050",
                    "fontFamily": "System",
                    "fontSize": 10.905714285714286,
                    "fontWeight": "bold",
                    "marginTop": 2.974285714285714,
                  }
                }
              >
                NEW MESSAGE - TAP TO OPEN
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "alignItems": "flex-end",
                "height": "100%",
                "justifyContent": "flex-start",
                "width": 63.885714285714286,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#505050",
                  "fontFamily": "System",
                  "fontSize": 9.914285714285715,
                  "fontStyle": "italic",
                  "fontWeight": "500",
                  "marginRight": 7.9314285714285715,
                  "marginTop": 7.9314285714285715,
                }
              }
            >
              Just now
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
